### PR TITLE
execute commands from the non-global environment

### PR DESCRIPTION
**Pull request by fatonsopa**
added: added the opportunity to execute commands from the non-global environment
Useful in case your commands are not accessible by global environment. ex. /usr/bin/local

### CODE REVIEW

**Comment by fatonsopa** [comment]
Hi,

Not trying to make you pressure :), but we are deploying a project and need to use your package with the binary path issue fixed. Would you please consider approving the PR (including laravel package) so we don't have to use a fork of the package?

Thank you in advance,
Faton

**Comment by fatonsopa** [comment]
Hi Spatie,

Any chance of approving this PR, so we can use the official package instead of the fork?

Thanks,
Faton

File: src/Optimizers/BaseOptimizer.php [hunk 1]
```diff
@@ -2,19 +2,81 @@
...
+      '/usr/bin',
+      '/usr/sbin',
+      '/usr/local/bin',
+      '/usr/local/sbin',
+      '/bin',
+      '/sbin'
+    ];
+
+    /** @var \Psr\Log\LoggerInterface */
+    protected $logger;
```
Comment (freekmurze) [inline src/Optimizers/BaseOptimizer.php]: Logging can already be done on the chain, remove everything regarding logging.

File: src/Optimizers/BaseOptimizer.php [hunk 2]
```diff
@@ -2,19 +2,81 @@
...
 {
     public $options = [];
 
     public $imagePath = '';
 
+    /**
+     * List of binary paths to check for commands.
+     * @var array $binaryPathList
+     */
+    protected $binaryPathList = [
```
Comment (freekmurze) [inline src/Optimizers/BaseOptimizer.php]: I don't want to maintain this list. Please make this an empty array instead.

File: src/Optimizers/BaseOptimizer.php [hunk 3]
```diff
@@ -2,19 +2,81 @@
...
+    ];
+
+    /** @var \Psr\Log\LoggerInterface */
+    protected $logger;
+
+    /**
+     * Binary name of the optimizer.
+     * @var string $binaryName
+     */
+    protected $binaryName;
```
Comment (freekmurze) [inline src/Optimizers/BaseOptimizer.php]: This should already be defined on a concrete optimizer

File: src/Optimizers/BaseOptimizer.php [hunk 4]
```diff
@@ -2,19 +2,81 @@
...
+
+    /**
+     * Set binary Path
+     *
+     * Useful in case your commands are not accessible by global environment. ex. /usr/bin/local
+     *
+     * @param string $binaryName
+     * @return string
+     */
+    public function setBinaryPath(string $binaryPath)
```
Comment (freekmurze) [inline src/Optimizers/BaseOptimizer.php]: make this accept both a string and an array

File: src/Optimizers/BaseOptimizer.php [hunk 5]
```diff
@@ -34,10 +96,66 @@ public function setOptions(array $options = [])
...
+        return $this->binaryPathList;
+    }
+
+    /**
+     * Authomatically detect the path where the image optimizes is installed
+     *
+     * @return $this
+     */
+    public function detectBinaryPath(){
+        // first check if comman is executed in a global environment
```
Comment (freekmurze) [inline src/Optimizers/BaseOptimizer.php]: This seems far to heavy. Can't we just check if the binary exists at the given path.

**Comment by fatonsopa** [comment]
Just committed requested changes. Can you please check?

Thanks,
Faton

**Comment by freekmurze** [comment]
For simplicity's sake I've decided to leave things as they are for now.

**Pull request closed without merging**